File: gcc/testsuite/gcc.dg/loop-versioning-8.c
/* { dg-options "-O3 -fdump-tree-lversion-details" } */

/* Versioning for step == 1 in these loops would allow loop interchange,
   but otherwise isn't worthwhile.  At the moment we decide not to version.  */

struct foo {
  int a[100];
};

void
f1 (struct foo *x, int step, int n)
{
  for (int i = 0; i < n; ++i)
    for (int j = 0; j < n; ++j)
      x[j * step].a[i] = 100;
}

void
f2 (struct foo *x, int step, int n)
{
  for (int i = 0; i < n; ++i)
    for (int j = 0; j < n; ++j)
      x[j].a[i * step] = 100;
}

void
f3 (struct foo *x, int step, int limit)
{
  for (int i = 0; i < 100; ++i)
    for (int j = 0; j < limit; j += step)
      x[j].a[i] = 100;
}

void
f4 (struct foo *x, int step, int limit)
{
  for (int i = 0; i < limit; i += step)
    for (int j = 0; j < 100; ++j)
      x[j].a[i] = 100;
}

/* { dg-final { scan-tree-dump-not {want to version} "lversion" } } */
/* { dg-final { scan-tree-dump-not {versioned} "lversion" } } */

File: gcc/testsuite/gfortran.dg/loop_versioning_8.f90
! { dg-options "-O3 -fdump-tree-lversion-details" }

! Check that versioning is applied to a gather-like reduction operation.

function f(x, index, n)
  integer :: n
  real :: x(:)
  integer :: index(n)
  f = sum(x(index(:)))
end function f

! { dg-final { scan-tree-dump-times {want to version containing loop} 1 "lversion" } }
! { dg-final { scan-tree-dump-times {versioned this loop} 1 "lversion" } }
